### ISSUE TITLE
Access to gradient of loss function with respect to input embedding

### ISSUE THREAD

**Issue by JJumSSu** [pytorch/fairseq#1216]
Hi, first of all, thanks for the great library you guys are providing.

I'm currently trying to implement an adversarial training scheme with fairseq library. In order to do so, I need to get the gradient vectors of loss function with respect to input sequence's embedding vectors every train step(batchwise). I'm currently using ```model.encoder.embed_tokens.weight``` as an embedding weight matrix and ```model.encoder.embed_tokens.weight.grad``` as a gradient of an embedding matrix. 

So every train step, I call this ```model.encoder.embed_tokens.weight.grad``` and get the indices of the gradient vectors with my input sequence tokens that are in the ```sample['net_input']['src_tokens']```. I was wondering, is this the right way to get the gradient of loss function with respect to input embeddings? Because in some cases during train step, the indexed gradient vectors are zero (not very often though)

Thank you!
